File: src/leetcode/python/medium/LongestAbsoluteFilePath_388.py
class LongestAbsoluteFilePath_388(object):
    def lengthLongestPath(self, input):
        """
        :type input: str
        :rtype: int
        """
        splits = input.split("\n")
        visited = [False] * len(input)
        tab = "\t"
        max_l = 0
        for i in range(len(splits)):
            if tab not in splits[i]:
                if "." in splits[i]:
                    max_l = max(max_l, len(splits[i]))
                    continue
                max_l = max(max_l, self.maxLength(splits, i+1, tab, len(splits[i]), visited))
        return max_l

    def maxLength(self, splits, start, tab, length, visited):
        max_l = 0
        i = start
        while i < len(splits):
            if tab in splits[i] and not visited[i]:
                tmp = len(splits[i]) - len(tab) + 1
                visited[i] = True
                if "." not in splits[i]:
                    max_l = max(self.maxLength(splits, i+1, (tab+"\t"), (length+tmp), visited), max_l)
                else:
                    max_l = max(length+tmp, max_l)
            elif not visited[i]:
                break
            i += 1
        return max_l

File: src/leetcode/java/LongestAbsoluteFilePath_388.java
package leetcode.java;

public class LongestAbsoluteFilePath_388 {
  public int lengthLongestPath(String input) {
    String[] splits = input.split("\n");
    boolean[] visited = new boolean[splits.length];
    String tab = "\t";
    int maxL = 0;
    for (int i = 0; i < splits.length; i++) {
      if (!splits[i].contains(tab)) {
        if (splits[i].contains(".")) {
          maxL = Math.max(maxL, splits[i].length());
          continue;
        }
        maxL = Math.max(maxL, maxLength(splits, i + 1, tab, splits[i].length(), visited));
      }
    }
    return maxL;
  }

  private int maxLength(String[] splits, int start, String tab, int length, boolean[] visited) {
    int maxL = 0;
    int tmpL;
    for (int i = start; i < splits.length; i++) {
      if (splits[i].contains(tab) && !visited[i]) {
        tmpL = splits[i].length() - tab.length() + 1;
        visited[i] = true;
        if (!splits[i].contains(".")) maxL = Math.max(maxLength(splits, i + 1, tab + "\t", length + tmpL, visited), maxL);
        else maxL = Math.max(length + tmpL, maxL);
      } else if (!visited[i]) break;
    }
    return maxL;
  }
}
